### PR TITLE
Implement SignGenericMulti for extended AccountProtectingMultiSigner

### PR DESCRIPTION
**Pull request by Bez625**
Implement SignGenericMulti in the dirk wallet

### CODE REVIEW

File: account.go [hunk 1]
```diff
@@ -107,6 +107,16 @@ func (a *account) SignGeneric(ctx context.Context, data []byte, domain []byte) (
 	return sig, nil
 }
 
+// SignGenericMulti signs multiple generic data roots.
+func (a *account) SignGenericMulti(ctx context.Context, data []byte, domain []byte, accounts []e2wtypes.Account) ([]e2types.Signature, error) {
```
Comment (mcdee) [inline account.go:111]: This should allow multiple `data` roots, as per the comment i.e. `data` should be of type `[][]byte`.

File: distributedaccount.go [hunk 2]
```diff
@@ -140,6 +140,16 @@ func (a *distributedAccount) SignGeneric(ctx context.Context, data []byte, domai
 	return sig, nil
 }
 
+// SignGenericMulti signs multiple generic data roots.
+func (a *distributedAccount) SignGenericMulti(ctx context.Context, data []byte, domain []byte, accounts []e2wtypes.Account) ([]e2types.Signature, error) {
```
Comment (mcdee) [inline distributedaccount.go:144]: Same comment as in `account.go`

File: grpc.go [hunk 3]
```diff
@@ -338,6 +338,137 @@ func (a *distributedAccount) SignGRPC(ctx context.Context,
...
+	}
+	req := &pb.MultisignRequest{
+		Requests: make([]*pb.SignRequest, len(accounts)),
+	}
+
+	for i := range accounts {
+		assertedAccount, isAccount := accounts[i].(*account)
+		if !isAccount {
+			return nil, errors.New("account not of required type")
+		}
```
Comment (mcdee) [inline grpc.go:374]: Is this necessary, given that this has already been checked above?

File: grpc.go [hunk 4]
```diff
@@ -338,6 +338,137 @@ func (a *distributedAccount) SignGRPC(ctx context.Context,
...
+
+	sigs := make([]e2types.Signature, len(accounts))
+	for i, response := range resp.GetResponses() {
+		if response.GetState() == pb.ResponseState_FAILED {
+			return nil, errors.New("request to obtain signatures failed")
+		}
+		if response.GetState() == pb.ResponseState_DENIED {
+			return nil, errors.New("request to obtain signatures denied")
+		}
+		span.AddEvent("Obtained signature bytes")
```
Comment (mcdee) [inline grpc.go:407]: This is likely to be quite noisy; not sure that we need this.

File: grpc.go [hunk 5]
```diff
@@ -338,6 +338,137 @@ func (a *distributedAccount) SignGRPC(ctx context.Context,
...
+	req := &pb.MultisignRequest{
+		Requests: make([]*pb.SignRequest, len(accounts)),
+	}
+
+	thresholds := make([]uint32, len(accounts))
+	for i := range accounts {
+		assertedAccount, isAccount := accounts[i].(*distributedAccount)
+		if !isAccount {
+			return nil, errors.New("account not of required type")
+		}
```
Comment (mcdee) [inline grpc.go:453]: As per above.

File: grpc.go [hunk 6]
```diff
@@ -338,6 +338,137 @@ func (a *distributedAccount) SignGRPC(ctx context.Context,
...
+		req.Requests[i] = &pb.SignRequest{
+			Id:     &pb.SignRequest_Account{Account: fmt.Sprintf("%s/%s", assertedAccount.wallet.Name(), accounts[i].Name())},
+			Data:   root,
+			Domain: domain,
+		}
+	}
+
+	ctx, cancelFunc := context.WithTimeout(ctx, a.wallet.timeout)
+	defer cancelFunc()
+	sig, err := a.thresholdMultiSign(ctx, req, thresholds)
```
Comment (mcdee) [inline grpc.go:464]: Should call this `sigs` as it contains multiple signatures.

**Review by mcdee — CHANGES_REQUESTED** [review]
Various comments . We should also update the version of `go-eth2-wallet-types` so that it has the correct interface definition.

File: grpc.go (hunk 3 above)
Comment (Bez625) [reply grpc.go:370]: I removed the check above this as I didn't see the need to iterate over the accounts twice.

**Comment by mcdee** [comment]
Please can you add a mock for the new function in `mock/erroring.go`?

**Comment by Bez625** [comment]
> Please can you add a mock for the new function in `mock/erroring.go`?

Done

**Pull request merged**